File: Koshelev/655.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int getHeight(TreeNode *root)
    {
	    if (!root)
		    return 0;
	    return max(getHeight(root->left), getHeight(root->right)) + 1;
    }

    void DFS(TreeNode* root, vector<vector<string>> &res, int r, int c, int height)
    {
        if (!root)
            return;
        res[r][c] = to_string(root->val);
        DFS(root->left, res, r + 1, c - pow(2, height - r - 1), height);
        DFS(root->right, res, r + 1, c + pow(2, height - r - 1), height);
    }
    vector<vector<string>> printTree(TreeNode* root) {
        int height = getHeight(root) - 1;
        int m = height + 1;
        int n = pow(2, height + 1) - 1;
        vector<vector<string>> res(m, vector<string>(n, ""));
        DFS(root, res, 0, (n - 1) / 2, height);
        return res;
    }
};

File: 655.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def printTree(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[str]]
        """
        h = self.height_tree(root)
        w = 2 ** h - 1
        mass = []
        for i in range(h):
            mass.append([""] * w)

        def dfs(root, left, right, i, mass):
            if root is None:
                return
            temp = (right + left) // 2
            mass[i][temp] = str(root.val)
            i += 1
            dfs(root.left, left, temp - 1, i, mass)
            dfs(root.right, temp + 1, right, i, mass)
            return

        dfs(root, 0, w - 1, 0, mass)

        return mass

    def height_tree(self, root):
        if root is not None:
            right_height = self.height_tree(root.right)
            left_height = self.height_tree(root.left)
            return max(right_height, left_height) + 1
        return 0
